pkg/tagger/global.go

```diff
@@ -62,11 +62,20 @@ func Init() {
 
 // Tag queries the defaultTagger to get entity tags from cache or sources.
 // It can return tags at high cardinality (with tags about individual containers),
-// or at orchestrator cardinality (pod/task level)
+// or at orchestrator cardinality (pod/task level).
 func Tag(entity string, cardinality collectors.TagCardinality) ([]string, error) {
 	return defaultTagger.Tag(entity, cardinality)
 }
 
+// TagReadOnly queries the defaultTagger to get entity tags from cache or
+// sources.  It can return tags at high cardinality (with tags about individual
+// containers), or at orchestrator cardinality (pod/task level). No copy will
+// be made of the internal data, the caller gurantees that no modifications
+// will be made to the slice.
+func TagReadOnly(entity string, cardinality collectors.TagCardinality) ([]string, error) {
```

Comment:
Can be call `UnsafeTag()` instead because we expose internal tagger data structure to other component.

I was thinking on another way to avoid the allocation. What about creating a new `tags.Builder` util,  some kind of [strings.Builder](https://golang.org/pkg/strings/#Builder) to manage the generation of the final tags list (`[]string`). 

something like:

```golang
// package tags
type Builder interface {
	Add(*[]string)
	Render() []string
}

// package tagger
type Tagger interface {
	TagBuilder(entity string, cardinality collectors.TagCardinality, builder *tags.Builder) error
}
```
We can initialised the `Builder` with the tags received in the metric payload. and then pass the Builder to the `Tagger`

The advantages that I see with this pattern are:
* the strings slice from the tagger is not expose to the user. only the builder can access it, so it reduce possible futur mistake.
* the final strings slice that contains all the tags is generate only in once, and managed by the builder.
* Since we know the size of each slices, the builder can compute the total size before creating it. which means less allocation.

WDTY?

Reply:
Not sure I understand then idea. We need to extend each dogstatsd message with the tagger metrics. The idea is to avoid allocating a slice. So allocating a Builder per message would defeat the purpose.

RIght now we have a slice allocated per context (1metrics every 10s). So every time we receive a message we reuse it to the context.

But, we could pass that slice to the tagger so the tagger directly happens the tags to it instead. WDYT ?

Reply:
sorry if my explanation wasn't clear. We need to create one Builder for each metric. 

```golang
func findOriginTags(origin string, builder tags.Builder) {
	if origin != listeners.NoOrigin {
		if err := tagger.TagBuilder(origin, tagger.DogstatsdCardinality, builder); err != nil {
			log.Errorf(err.Error())
		} 
	}

	// Include orchestrator scope tags if the cardinality is set to orchestrator
	if tagger.DogstatsdCardinality == collectors.OrchestratorCardinality {
		 if err := tagger.TagBuilder(collectors.OrchestratorScopeEntityID, tagger.OrchestratorCardinality, builder); err != nil {
			log.Errorf(err.Error())
		} 
	}
}

Reply:
I update the PR, let me know what you think